src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.HillClimbing.cs

```diff
@@ -0,0 +1,385 @@
+
+                //
+                // Record these numbers for posterity
+                //
+                
+                // Worker Thread Adjustment stats event
+
+
+                //
+                // If all of this caused an actual change in thread count, log that as well.
+                // 
+                if (newThreadCount != currentThreadCount)
+                    ChangeThreadCount(newThreadCount, state);
+
+                //
+                // Return the new thread count and sample interval.  This is randomized to prevent correlations with other periodic
+                // changes in throughput.  Among other things, this prevents us from getting confused by Hill Climbing instances
+                // running in other processes.
+                //
+                // If we're at minThreads, and we seem to be hurting performance by going higher, we can't go any lower to fix this.  So
+                // we'll simply stay at minThreads much longer, and only occasionally try a higher value.
+                //
+                int newSampleInterval;
+                if (ratio.real < 0.0 && newThreadCount == minThreads)
+                    newSampleInterval = (int)(0.5 + _currentSampleInterval * (10.0 * Math.Max(-ratio.real, 1.0)));
+                else
+                    newSampleInterval = _currentSampleInterval;
+
+                return (newThreadCount, newSampleInterval);
+            }
+
+            private void ChangeThreadCount(int newThreadCount, State state)
+            {
+                _lastThreadCount = newThreadCount;
+                _currentSampleInterval = _randomIntervalGenerator.Next(_sampleIntervalLow, _sampleIntervalHigh);
+                double throughput = _secondsElapsedSinceLastChange > 0 ? _completionsSinceLastChange / _secondsElapsedSinceLastChange : 0;
+                LogTransition(newThreadCount, throughput, state);
+                _secondsElapsedSinceLastChange = 0;
+                _completionsSinceLastChange = 0;
+            }
+
+            private void LogTransition(int newThreadCount, double throughput, State state)
+            {
+                // TODO: Log transitions
+            }
+
+            public void ForceChange(int newThreadCount, State state)
+            {
+                if(_lastThreadCount != newThreadCount)
+                {
+                    _currentControlSetting += newThreadCount - _lastThreadCount;
+                    ChangeThreadCount(newThreadCount, state);
+                }
+            }
+
+            private static (double real, double imaginary) GetWaveComponent(double[] samples, int numSamples, double period)
+            {
+                double w = 2 * Math.PI / period;
+                double cos = Math.Cos(w);
+                double coeff = 2 * cos;
+                double q0 = 0, q1 = 0, q2 = 0;
+                for(int i = 0; i < numSamples && i < samples.Length; ++i)
```

Comment:
Could just have `i < numSamples` as the loop condition and assert in the body that `i < samples.Length`. An invalid `i` would trigger an assert that would be easier to debug than an exception, and we would get an exception from the inherent bound check anyway.